File: Sem2 MAP/src/InMemoryRepository.java
import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID,E> {

    protected Map<ID,E> entities;

    public InMemoryRepository() {
        entities=new HashMap<ID,E>();
    }

    @Override
    public E findOne(ID id) {
        if(id==null)
            throw new IllegalArgumentException("id is null");
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        /**
         *
         * @param entity
         *         entity must be not null
         * @return null- if the given entity is saved
         *         otherwise returns the entity (id already exists)
         * @throws ValidationException
         *            if the entity is not valid
         * @throws IllegalArgumentException
         *             if the given entity is null.     *
         */
        if(entity==null)
            throw new IllegalArgumentException("ENTITY CANNOT BE NULL");
        if(entities.containsKey(entity.getId()))
            return entity;
        else{
            entities.put(entity.getId(),entity);
            return null;
        }
    }

    @Override
    public E delete(ID id) {
        if(id==null)
            throw new IllegalArgumentException("id is null");
        if(!entities.containsKey(id)){
            return null;
        }
        else {
            return entities.remove(id);
        }
    }

    @Override
    public E update(E entity) {
        if(entity==null)
            throw new IllegalArgumentException("ENTITY CANNOT BE NULL");
        if(!entities.containsKey(entity.getId())){
            entities.put(entity.getId(),entity);
            return entity;
        }
        else{
            entities.put(entity.getId(),entity);
            return null;
        }
    }
}

File: Sem10-MAP/Sem10-MAP/Repository/InMemoryRepository.cs
using Sem10_MAP.Domain;

namespace Sem10_MAP.Repository;

public class InMemoryRepository<E, ID>: IRepository<E, ID> where E: Entity<ID>
{
    Dictionary<ID, E> Dictionary = new();
    
    public E? FindOne(ID id)
    {
        Dictionary.TryGetValue(id, out E? value);
        return value;
    }
    
    public IEnumerable<E> FindAll()
    {
        return Dictionary.Values;
    }

    public virtual E? Save(E entity)
    {
        return Dictionary.TryAdd(entity.Id, entity) ? entity : null;
    }
    
    public virtual E? Delete(ID id)
    {
        Dictionary.Remove(id, out E? entity);
        return entity;
    }

    public virtual E? Update(E entity)
    {
        if (!Dictionary.ContainsKey(entity.Id))
        {
            return null;
        }
        Dictionary[entity.Id] = entity;
        return entity;
    }
}